### PR TITLE
Feature/nft subgraph

### PR DESCRIPTION
**Pull request by novaknole**
## Description

Please include a summary of the change and be sure you follow the contributions rules we do provide [here](./CONTRIBUTIONS.md)

Task: [ID]()

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the test network.

### CODE REVIEW

File: packages/subgraph/schema.graphql [hunk 1]
```diff
@@ -8,29 +8,105 @@ interface Token {
...
+type ERC20Balance implements TokenBalance @entity {
+  id: ID! # token address + dao address
+  token: ERC20Contract! # address(0) corresponds to the native token
   dao: Dao!
   balance: BigInt!
   lastUpdated: BigInt!
 }
 
-# Dao
+type ETHBalance implements TokenBalance @entity {
```
Comment (Michael-A-Heuer) [inline packages/subgraph/schema.graphql:39]: Better use `NativeBalance` as we will also be on Polygon and other chains/layers where the native token is not ETH

File: packages/subgraph/schema.graphql [hunk 2]
```diff
@@ -8,29 +8,105 @@ interface Token {
...
+  tokenId: BigInt!
+  from: Bytes!
+  to: Bytes!
+  proposal: Proposal
+  type: TransferType!
+  txHash: Bytes!
+  createdAt: BigInt!
+}
+
+type ETHTransfer implements TokenTransfer @entity(immutable: true) {
```
Comment (Michael-A-Heuer) [inline packages/subgraph/schema.graphql:98]: `NativeTransfer`

File: packages/subgraph/schema.graphql (hunk 1 above)
Comment (novaknole) [reply packages/subgraph/schema.graphql]: done https://github.com/aragon/core/pull/251/commits/72d501bfd5d066a003a534043150f5f6667b75b4

File: packages/subgraph/schema.graphql (hunk 2 above)
Comment (novaknole) [reply packages/subgraph/schema.graphql]: https://github.com/aragon/core/pull/251/commits/72d501bfd5d066a003a534043150f5f6667b75b4

File: packages/subgraph/schema.graphql [hunk 3]
```diff
@@ -101,10 +157,10 @@ type Dao @entity {
   creator: Bytes!
   metadata: String
   createdAt: BigInt!
-  token: ERC20Token
+  token: ERC20Contract # TODO
```
Comment (brickpop) [inline packages/subgraph/schema.graphql:160]: TO DO

File: packages/subgraph/src/packages/admin/admin.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import {dataSource, DataSourceContext} from '@graphprotocol/graph-ts';
+import {dataSource, DataSourceContext, log} from '@graphprotocol/graph-ts';
```
Comment (brickpop) [inline packages/subgraph/src/packages/admin/admin.ts:1]: Used?

File: packages/subgraph/src/utils/tokens/erc20.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  erc20Transfer.txHash = event.transaction.hash;
+  erc20Transfer.createdAt = event.block.timestamp;
+  erc20Transfer.token = contract.id;
+  erc20Transfer.proposal = proposalId;
+
+  // If from/to both aren't equal to dao, it means
+  // dao must have been approved for the `amount`
+  // and played the role of transfering between 2 parties.
+  if (from != dao && to != dao) {
+    erc20Transfer.type = 'None'; // No idea
```
Comment (Michael-A-Heuer) [inline packages/subgraph/src/utils/tokens/erc20.ts:148]: Wouldn't this mean the DAO has sent money on behalf of someone else that approved the DAO as a spender?

Not sure if we should still store it as the DAO acted / did something

File: packages/subgraph/src/utils/tokens/erc721.ts [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+  let tuple = decoded.toTuple();
+
+  let from = tuple[0].toAddress();
+  let to = tuple[1].toAddress();
+  let tokenId = tuple[2].toBigInt();
+
+  let daoId = dao.toHexString();
+
+  // Ambiguity ! No need to store transfer such as this.
```
Comment (Michael-A-Heuer) [inline packages/subgraph/src/utils/tokens/erc721.ts:178]: Same as before

File: packages/subgraph/src/utils/tokens/erc721.ts [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+  transfer.proposal = proposalId;
+  transfer.txHash = event.transaction.hash;
+  transfer.createdAt = event.block.timestamp;
+  transfer.save();
+
+  // If from/to both aren't equal to dao, it means
+  // dao must have been approved for the `tokenId`
+  // and played the role of transfering between 2 parties.
+  if (from != dao && to != dao) {
+    transfer.type = 'None'; // No idea
```
Comment (Michael-A-Heuer) [inline packages/subgraph/src/utils/tokens/erc721.ts:204]: Same as before

File: packages/subgraph/src/utils/tokens/erc20.ts [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+  }
+
+  if (functionSelector == ERC20_transferFrom) {
+    from = tuple[0].toAddress();
+    to = tuple[1].toAddress();
+    amount = tuple[2].toBigInt();
+  }
+
+  // Ambiguity ! No need to store transfer such as this.s
+  if (from == to) {
```
Comment (Michael-A-Heuer) [inline packages/subgraph/src/utils/tokens/erc20.ts:121]: Why is this ambiguous? 

`from == to` makes no sense ofc, but I am not sure if we shouldn't handle these edge cases differently than simply returning.

Isn't this, formally, still a tx / a transfer that we should index?

File: packages/subgraph/src/utils/tokens/eth.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+export function handleNativeAction(
+  dao: Address,
+  to: Address,
+  amount: BigInt,
+  reference: string,
+  proposalId: string,
+  actionIndex: number,
+  event: ethereum.Event
+): void {
+  if (dao == to) {
```
Comment (Michael-A-Heuer) [inline packages/subgraph/src/utils/tokens/eth.ts:69]: Same as before

File: packages/subgraph/tests/dao/dao.test.ts [hunk 10]
```diff
@@ -52,6 +55,31 @@ test('Run dao (handleMetadataSet) mappings with mock event', () => {
...
+  //   DAO_ADDRESS,
+  //   Bytes.fromHexString('0x150b7a02'),
+  //   ADDRESS_ONE, // ToKen Address
+  //   encoded
+  // )
+  // handleCallbackReceived(newEvent);
+  // // // let d = ethereum.Tuple.from(tupleArray);
+  // // let decoded = ethereum.decode("(address,address,uint256,bytes)", encoded)!.toTuple();
+  // // // let encoded = ethereum.encode(ethereum.Value.fromTuple(d));
+  // // log.warning("what {}", [encoded.toHexString()])
```
Comment (Michael-A-Heuer) [inline packages/subgraph/tests/dao/dao.test.ts:80]: Do we have a follow-up task to add these tests?

File: packages/subgraph/tests/dao/dao.test.ts [hunk 11]
```diff
@@ -60,251 +88,245 @@ test('Run dao (handleNativeTokenDeposited) for native token mappings with mock e
...
-  assert.fieldEquals(
-    'VaultTransfer',
-    entityID,
-    'createdAt',
-    event.block.timestamp.toString()
-  );
-
-  clearStore();
-});
+// test('Run dao (handleDeposited) for Token mappings with mock event', () => {
```
Comment (Michael-A-Heuer) [inline packages/subgraph/tests/dao/dao.test.ts:130]: Do we have a follow-up task to add these tests?

**Review by Michael-A-Heuer — COMMENTED** [review]
I have some uncertainty about how we handle edge cases such as:

- the DAO being the spender of a transaction with external `from` and `to` address for (ERC20, ERC721)
- the DAO sending tokens (Native, ERC20, ERC721) to itself.

If this was a proposal that went through and got executed, don't we want to still be able to show these txns in the finance section? 
For example, there could be an admin who accidentally sent tokens with `from == to`. If it shows up in the finance sections, this would allow him to notice.

File: packages/subgraph/src/packages/admin/admin.ts (hunk 4 above)
Comment (novaknole) [reply packages/subgraph/src/packages/admin/admin.ts:1]: done

File: packages/subgraph/schema.graphql (hunk 3 above)
Comment (novaknole) [reply packages/subgraph/schema.graphql:160]: removed

[the remark this answers is not in the document]
Comment (novaknole) [reply packages/subgraph/schema.graphql:160]: removed todo, but will look into it later on the next PR.

File: packages/subgraph/src/dao/dao.ts [hunk 12]
```diff
@@ -37,153 +47,100 @@ export function _handleMetadataSet(daoId: string, metadata: string): void {
...
+}
+
+// ERC20 + Native
+export function handleDeposited(event: Deposited): void {
+  if (event.params.token.toHexString() != ADDRESS_ZERO) {
+    handleERC20Deposit(
+      event.address,
+      event.params.token,
+      event.params.sender,
+      // event.params._reference, TODO: pass it
```
[the remark this answers is not in the document]
Comment (novaknole) [reply packages/subgraph/src/dao/dao.ts:121]: removed. might decide to do it in future pr.

File: packages/subgraph/tests/dao/dao.test.ts (hunk 10 above)
Comment (novaknole) [reply packages/subgraph/tests/dao/dao.test.ts:80]: Yes, starting to work on this now..

File: packages/subgraph/tests/dao/dao.test.ts (hunk 11 above)
Comment (novaknole) [reply packages/subgraph/tests/dao/dao.test.ts:130]: yep. working on it.

File: packages/subgraph/tests/dao/dao.test.ts [hunk 13]
```diff
@@ -52,258 +135,698 @@ test('Run dao (handleMetadataSet) mappings with mock event', () => {
...
+    let balanceId = daoId.concat('_').concat(tokenId);
+    let transferId = getTransferId(txHash, logIndex, 0);
+
+    // check NativeBalance entity
+    eq('NativeBalance', balanceId, 'id', balanceId);
+    eq('NativeBalance', balanceId, 'dao', daoId);
+    eq('NativeBalance', balanceId, 'balance', ONE_ETH);
+    eq('NativeBalance', balanceId, 'lastUpdated', timestamp.toString());
+
+    // Check ERC20Transfer
```
Comment (Michael-A-Heuer) [inline packages/subgraph/tests/dao/dao.test.ts:308]:
```suggestion
    // Check NativeTransfer
```

File: packages/subgraph/tests/dao/dao.test.ts [hunk 14]
```diff
@@ -52,258 +135,698 @@ test('Run dao (handleMetadataSet) mappings with mock event', () => {
...
+
+        // check ERC20Balance entity
+        eq('ERC20Balance', balanceId, 'id', balanceId);
+        eq('ERC20Balance', balanceId, 'token', tokenId);
+        eq('ERC20Balance', balanceId, 'dao', daoId);
+        eq('ERC20Balance', balanceId, 'balance', ERC20_AMOUNT_HALF);
+        eq('ERC20Balance', balanceId, 'lastUpdated', timestamp.toString());
+        assert.entityCount('ERC20Balance', 1);
+
+        // Check ERC721Transfer
```
Comment (Michael-A-Heuer) [inline packages/subgraph/tests/dao/dao.test.ts:524]:
```suggestion
        // Check ERC20Transfer
```

File: packages/subgraph/tests/dao/dao.test.ts [hunk 15]
```diff
@@ -52,258 +135,698 @@ test('Run dao (handleMetadataSet) mappings with mock event', () => {
...
```
Comment (Michael-A-Heuer) [inline packages/subgraph/tests/dao/dao.test.ts:604]:
```suggestion
        // Check ERC20Transfer
```

File: packages/subgraph/tests/dao/dao.test.ts [hunk 16]
```diff
@@ -52,258 +135,698 @@ test('Run dao (handleMetadataSet) mappings with mock event', () => {
...
+        eq('ERC20Balance', balanceId, 'token', tokenId);
+        eq('ERC20Balance', balanceId, 'dao', daoId);
+        eq('ERC20Balance', balanceId, 'balance', ERC20_AMOUNT_HALF);
+        eq('ERC20Balance', balanceId, 'lastUpdated', timestamp.toString());
+        assert.entityCount('ERC20Balance', 1);
+
+        // Check ERC20Transfer
+        eq('ERC20Transfer', transferId, 'id', transferId);
+        eq('ERC20Transfer', transferId, 'dao', daoId);
+        eq('ERC20Transfer', transferId, 'amount', '10'); // TODO: fix it 10
```
Comment (Michael-A-Heuer) [inline packages/subgraph/tests/dao/dao.test.ts:527]: What needs to be fixed here?

File: packages/subgraph/tests/dao/dao.test.ts [hunk 17]
```diff
@@ -52,258 +135,698 @@ test('Run dao (handleMetadataSet) mappings with mock event', () => {
...
+
+        handleExecuted(event);
+
+        let txHash = event.transaction.hash;
+        let logIndex = event.transactionLogIndex;
+        let timestamp = event.block.timestamp;
+
+        let transferId = getTransferId(txHash, logIndex, 0);
+
+        // check ERC720Contract entity
```
Comment (Michael-A-Heuer) [inline packages/subgraph/tests/dao/dao.test.ts:590]:
```suggestion
        // check ERC20Contract entity
```

File: packages/subgraph/tests/dao/dao.test.ts [hunk 18]
```diff
@@ -52,258 +135,698 @@ test('Run dao (handleMetadataSet) mappings with mock event', () => {
...
+        eq('ERC20Balance', balanceId, 'token', tokenId);
+        eq('ERC20Balance', balanceId, 'dao', daoId);
+        eq('ERC20Balance', balanceId, 'balance', ERC20_AMOUNT_HALF);
+        eq('ERC20Balance', balanceId, 'lastUpdated', timestamp.toString());
+        assert.entityCount('ERC20Balance', 1);
+
+        // Check ERC20Transfer
+        eq('ERC20Transfer', transferId, 'id', transferId);
+        eq('ERC20Transfer', transferId, 'dao', daoId);
+        eq('ERC20Transfer', transferId, 'amount', '10'); // TODO: fix it 10
```
Comment (Michael-A-Heuer) [inline packages/subgraph/tests/dao/dao.test.ts:607]: What needs to be fixed here?

Comment (novaknole) [reply packages/subgraph/tests/dao/dao.test.ts:607]: pushed. no worries.

File: packages/subgraph/schema.graphql [hunk 19]
```diff
@@ -8,29 +8,106 @@ interface Token {
...
+  token: ERC721Contract!
+  dao: Dao!
+  tokenIds: [BigInt!]!
+  lastUpdated: BigInt!
+}
+
+# Token Transfers
+
+enum TransferType {
+  None # Special case where from and to both are NOT equal to dao address.
```
[the remark this answers is not in the document]
Comment (novaknole) [reply packages/subgraph/schema.graphql]: yep. done.

File: packages/subgraph/tests/dao/dao.test.ts [hunk 20]
```diff
@@ -492,9 +492,10 @@ describe('handleExecuted', () => {
 
     describe('ERC20 transfer action', () => {
       test('creates entities with correct values', () => {
+        let transferToken = BigInt.fromU32(10);
```
Comment (Michael-A-Heuer) [inline packages/subgraph/tests/dao/dao.test.ts:500]: I would have called this amount. However, I can approve

**Review by Michael-A-Heuer — DISMISSED** [review]
Work will be continued, so good to go to unblock others

File: packages/subgraph/scripts/deploy-subgraph.sh [hunk 21]
```diff
@@ -45,8 +45,8 @@ then
         --node http://localhost:8020
 else
     graph deploy $FULLNAME \
-        --version-label $SUBGRAPH_VERSION \
-        --node https://app.satsuma.xyz/api/subgraphs/deploy \
+        --ipfs https://api.thegraph.com/ipfs/ \
```
[the remark this answers is not in the document]
Comment (mathewmeconry) [reply packages/subgraph/scripts/deploy-subgraph.sh:48]: just use a  `.env`

File: packages/subgraph/scripts/deploy-subgraph.sh [hunk 22]
```diff
@@ -45,8 +45,8 @@ then
...
+        --node https://api.thegraph.com/deploy/ \
```
Comment (mathewmeconry) [inline packages/subgraph/scripts/deploy-subgraph.sh:49]: revert that back to satsuma before merge

File: packages/subgraph/src/utils/tokens/erc20.ts [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+
+  contract.name = try_name.reverted ? '' : try_name.value;
+  contract.symbol = try_symbol.reverted ? '' : try_symbol.value;
+  contract.decimals = try_decimals.reverted ? 18 : try_decimals.value;
+  contract.save();
+
+  return contract;
+}
+
+export function updateERC20Balance(
```
Comment (mathewmeconry) [inline packages/subgraph/src/utils/tokens/erc20.ts:45]: same here. skip checks if an entity for this token is already present

File: packages/subgraph/src/utils/tokens/erc721.ts [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+  dao: Address,
+  data: Bytes,
+  event: ethereum.Event
+): void {
+  let contract = fetchERC721(token);
+  if (!contract) {
+    return;
+  }
+
+  let calldata = DECODE_OFFSET + data.toHexString().slice(10); // TODO: Giorgi insted aof 0x, append DECODE_OFFSET
```
Comment (mathewmeconry) [inline packages/subgraph/src/utils/tokens/erc721.ts:93]: do the TODO

**Review by mathewmeconry — CHANGES_REQUESTED** [review]

File: packages/subgraph/schema.graphql [hunk 25]
```diff
@@ -8,29 +8,106 @@ interface Token {
...
+  token: ERC721Contract!
+  dao: Dao!
+  tokenIds: [BigInt!]!
+  lastUpdated: BigInt!
+}
+
+# Token Transfers
+
+enum TransferType {
+  ExternalTransfer # when from/to are both NOT equal to dao.
```
[the remark this answers is not in the document]
Comment (mathewmeconry) [reply packages/subgraph/schema.graphql:57]: because they don't change our balance and thus don't touch the treasury

File: packages/subgraph/schema.graphql [hunk 26]
```diff
@@ -8,29 +8,106 @@ interface Token {
...
+  balance: BigInt!
+  lastUpdated: BigInt!
+}
+
+type ERC721Balance implements TokenBalance @entity {
+  id: ID!
+  token: ERC721Contract!
+  dao: Dao!
+  tokenIds: [BigInt!]!
+  lastUpdated: BigInt!
```
[the remark this answers is not in the document]
Comment (novaknole) [reply packages/subgraph/schema.graphql:51]: 100%. I know(I got this on my roadmap, but this PR took my energy and amazing tests, so have to give it up for a while to stay sane..

File: packages/subgraph/schema.graphql (hunk 25 above)
[the remark this answers is not in the document]
Comment (novaknole) [reply packages/subgraph/schema.graphql:57]: The alignment was `yes`. So really don't care !

File: packages/subgraph/src/utils/tokens/erc20.ts [hunk 27]
```diff
@@ -0,0 +1,199 @@
...
+} from '../../../generated/schema';
+import {ERC20} from '../../../generated/templates/DaoTemplate/ERC20';
+import {
+  ERC20_transfer,
+  ERC20_transferFrom,
+  getTransferId,
+  TransferType
+} from './common';
+
+export function fetchERC20(address: Address): ERC20Contract | null {
```
[the remark this answers is not in the document]
Comment (mathewmeconry) [reply packages/subgraph/src/utils/tokens/erc20.ts:15]: dumb me. you are doing that already

File: packages/subgraph/src/utils/tokens/erc721.ts (hunk 24 above)
Comment (novaknole) [reply packages/subgraph/src/utils/tokens/erc721.ts]: There was a comment remaining, it's actually done. So I removed the TODO comment.

File: packages/subgraph/scripts/deploy-subgraph.sh (hunk 22 above)
Comment (novaknole) [reply packages/subgraph/scripts/deploy-subgraph.sh]: done

File: packages/subgraph/scripts/deploy-subgraph.sh [hunk 28]
```diff
@@ -30,7 +30,7 @@ then
 fi
 
 # Prepare subgraph name
-FULLNAME=$SUBGRAPH_NAME-$NETWORK_NAME
+FULLNAME=novaknole/core-test-1
```
[the remark this answers is not in the document]
Comment (novaknole) [reply packages/subgraph/scripts/deploy-subgraph.sh]: done

**Review by Michael-A-Heuer — APPROVED** [review]
